src/compile/typecheck.rs

```diff
@@ -291,6 +292,255 @@ fn inline(
     }
 }
 
+///Discovers code segments that could never be executed
+fn flowcheck_reachability<T: AbstractSyntaxTree>(
+    node: &mut T,
+    file_info_chart: &BTreeMap<u64, FileInfo>,
+) {
+    let mut children = node.child_nodes();
+    let mut child_iter = children.iter_mut();
+
+    let mut locations = vec![];
+
+    for child in &mut child_iter {
+        match child {
+            TypeCheckedNode::Statement(stat) => match &mut stat.kind {
+                TypeCheckedStatementKind::Return(_) | TypeCheckedStatementKind::ReturnVoid() => {
+                    locations.extend(stat.debug_info.location);
+                    break;
+                }
+                TypeCheckedStatementKind::Expression(expr) => match &mut expr.kind {
+                    TypeCheckedExprKind::If(_, block, else_block, ..)
+                    | TypeCheckedExprKind::IfLet(_, _, block, else_block, ..) => {
+                        flowcheck_reachability(block, file_info_chart);
+
+                        if let Some(branch) = else_block {
+                            flowcheck_reachability(branch, file_info_chart);
+                        }
+
+                        continue;
+                    }
+                    _ => {}
+                },
+                _ => {}
+            },
+            _ => {}
+        }
+
+        flowcheck_reachability(child, file_info_chart);
+    }
+
+    match child_iter.next() {
+        Some(TypeCheckedNode::Statement(issue)) => locations.extend(issue.debug_info.location),
+        _ => {}
+    };
+
+    match child_iter.last() {
+        Some(TypeCheckedNode::Statement(issue)) => locations.extend(issue.debug_info.location),
+        _ => {}
+    };
+
+    if locations.len() <= 1 {
+        return;
+    }
+
+    CompileError::new(
+        String::from("Compile warning"),
+        if locations.len() == 2 {
+            String::from("found unreachable statement")
+        } else {
+            String::from("found unreachable statements")
+        },
+        locations,
+        true,
+    )
+    .warn(file_info_chart);
+}
+
+///Discovers assigned values that are never used
+fn flowcheck_liveliness(
+    mut nodes: Vec<TypeCheckedNode>,
+    problems: &mut Vec<(Location, StringId)>,
+    loop_pass: bool,
+) -> Vec<StringId> {
+    let mut node_iter = nodes.iter_mut();
+    let mut alive = BTreeMap::new(); // values that are alive and need to die in this scope
+    let mut born = BTreeSet::<StringId>::new(); // values brought to life within this scope
+    let mut killed = vec![]; // values this scope has killed
+
+    // algorithm notes:
+    //   anything still alive at the end of scope is unused
+    //   scopes return everything they've killed
+    //   we make the simplifying assumption that an assignment before the first let is always used later
+    //   loops are unrolled to give assignments a second chance to be killed before leaving scope
+    //   conditionals are assumed to be
```

Comment:
Looks like this comment is unfinished.

Reply:
Good catch: I've completed the comment and added additional information for further understanding about assumptions made.